### ISSUE TITLE
Performance drops upon sleep call

### ISSUE THREAD

**Issue by malloc2k** [ocornut/imgui#8249]
### Version/Branch of Dear ImGui:

Version 1.91.5, Branch: master

### Back-ends:

imgui_impl_dx11.cpp + imgui_impl_win32.cpp

### Compiler, OS:

Windows 11 + MSVC2022

### Full config/build information:

```
Dear ImGui 1.91.5 (19150)
--------------------------------

sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=199711
define: _WIN32
define: _WIN64
define: _MSC_VER=1942
define: _MSVC_LANG=202002
--------------------------------

io.BackendPlatformName: imgui_impl_win32
io.BackendRendererName: imgui_impl_dx11
io.ConfigFlags: 0x00000000
io.ConfigNavCaptureKeyboard
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000000E
 HasMouseCursors
 HasSetMousePos
 RendererHasVtxOffset
--------------------------------

io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,128
io.DisplaySize: 1920.00,1080.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------

style.WindowPadding: 15.00,15.00
style.WindowBorderSize: 1.00
style.FramePadding: 8.00,6.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 10.00,8.00
style.ItemInnerSpacing: 4.00,4.00
```

### Details:

**My Issue/Question:**

I have a loop where I call my app's functions that need to be called each frame. For CPU optimization I am trying to introduce a sleep of 1ms to the main loop in order to burn less resources, the problem is that if I introduce **ANY** sleep to the main loop (can be 1 microsecond) ImGui gets locked to around 64FPS for no aparent reason. I have tried this with both V-Sync enabled and disabled but the output is the same.

`std::this_thread::sleep_for(std::chrono::milliseconds(1));`

I'm using the "standard" sleep calls that C++ provides.

Any help would be greatly appreciated!

### Minimal, Complete and Verifiable Example code:

```cpp
// Example function thats called every frame until the app is closed

bool Window::Render()
{

//Process the messages
MSG message;
        while (PeekMessage(&message, nullptr, 0U, 0U, PM_REMOVE)) {
            TranslateMessage(&message);
            DispatchMessage(&message);
        }
        if (message.message == WM_QUIT)
            bRunning = false;
            
        //Draw the frames
        ImGui_ImplDX11_NewFrame();
        ImGui_ImplWin32_NewFrame();
        ImGui::NewFrame();
        
        
        //Call the ImGui renders 
        
        std::this_thread::sleep_for(std::chrono::milliseconds(1)); //When this sleep is added, ImGui takes 15.6 ms each frame for no reason?


return true;
        

}

```

**Comment by malloc2k** [comment]
I forgot to mention but I have 2 165Hz monitors and DX11 is configured to render at 165Hz

**Comment by ocornut** [comment]
> std::this_thread::sleep_for(std::chrono::milliseconds(1)); //When this sleep is added, ImGui takes 15.6 ms each frame for no reason?

The question in fact has nothing to do with imgui.
You probably have a swap/vsync wait somewhere that by default takes 15.6 ms.
If you remove the imgui calls you will likely get the same results.

> DX11 is configured to render at 165Hz

I am guessing this assertion is in fact incorrect.

**Comment by malloc2k** [comment]
> > std::this_thread::sleep_for(std::chrono::milliseconds(1)); //When this sleep is added, ImGui takes 15.6 ms each frame for no reason?
> 
> The question in fact has nothing to do with imgui. You probably have a swap/vsync wait somewhere that by default takes 15.6 ms. If you remove the imgui calls you will likely get the same results.
> 
> > DX11 is configured to render at 165Hz
> 
> I am guessing this assertion is in fact incorrect.

How come that when I add p.e. 1 microsecond of sleep to my main loop I always get 64FPS on imgui? I don’t think I have anything in my code that would make it take exactly 15.6ms to swap. 

As per the DX11 I meant that I’ve set the swap chain refresh rate to 165. 

Any clues?

**Comment by ocornut** [comment]
> How come that when I add p.e. 1 microsecond of sleep to my main loop I always get 64FPS on imgui? I don’t think I have anything in my code that would make it take exactly 15.6ms to swap.

But I'm telling you if you remove every imgui code you will get the same result and therefore this is not an imgui question but a DX11/Windows question.

The clue that I am challenging your "I’ve set the swap chain refresh rate to 165." statement. Your SwapBuffer call is likely taking 15.6 ms which is exactly what's left at 60 FPS when removing 1 ms. Graphics API, swap chains and drivers are complicated and don't always do what you expect.

**Comment by malloc2k** [comment]
> > How come that when I add p.e. 1 microsecond of sleep to my main loop I always get 64FPS on imgui? I don’t think I have anything in my code that would make it take exactly 15.6ms to swap.
> 
> But I'm telling you if you remove every imgui code you will get the same result and therefore this is not an imgui question but a DX11/Windows question.
> 
> The clue that I am challenging your "I’ve set the swap chain refresh rate to 165." statement. Your SwapBuffer call is likely taking 15.6 ms which is exactly what's left at 60 FPS when removing 1 ms. Graphics API, swap chains and drivers are complicated and don't always do what you expect.
> 
> I'm also not sure sure what precision guarantee `std::this_thread::sleep_for` provides. But try to comment all imgui code first.

I will comment it out later today and get back to you. Thanks 🙏

**Comment by malloc2k** [comment]
Hey @ocornut so I ended up figuring out what was happening. 

Windows resolution timer is at around 15.6ms which makes any sleep call (at least in my machine) below that value round up to it, rendering any call below ineffective and also blocking. I ended up using Multimedia resolution timers to set it down to 1ms. If anyone ever has this issue here's a link [https://learn.microsoft.com/en-us/windows/win32/multimedia/obtaining-and-setting-timer-resolution](url)
